### PR TITLE
Add cross-validation report for classification

### PR DESCRIPTION
**Pull request by dccastro**
Creates a new report template for cross-validation results (currently only for binary classification), which is hooked into `InnerEye.ML.visualizers.plot_cross_validation.plot_cross_validation_from_files()`. It plots model output statistics and ROC and PR curves for all crossval folds, along with aggregates and confidence intervals.

**UPDATE:** This was completely overhauled since the PR was initially opened. There is now a single classification report template used by both single and cross-validation runs (`classification_crossval_report.ipynb`), and the dispatching happens inside the individual functions in `classification_report.py`, sharing as much code as possible. Currently, the cross-validation report contains tables of classification metrics and PR/ROC curves, and skips the sections with best and worst samples.

This PR also adds a mechanism to load metrics for a specific crossval split and Train/Val/Test split from a CSV, rather than always loading the entire file. This enables reading cross-validation metrics from `outputs/CrossValResults/MetricsAcrossAllRuns.csv`.

### CODE REVIEW

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "df_ = df.rename(columns={'subject': 'series'}).join(dataset[['subject', 'series', 'class_name']].set_index('series'), on='series', rsuffix='_')\n",
+    "# df.rename(columns={'subject': 'series'})['series']#.join(dataset, on='series')"
```
Comment (ant0nsc) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:91]: commented code should  be removed

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plt.hist([df[df.label==0].model_output, df[df.label==1].model_output],\n",
+    "#     plt.hist([logit(fold_df[fold_df.label==0].model_output), logit(fold_df[fold_df.label==1].model_output)],\n",
+    "#              density=True, bins=np.linspace(-6, 6, 21), histtype='step');\n",
```
Comment (ant0nsc) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:132]: commented out code?

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    "         density=True, bins=16, histtype='step', lw=2, label=['CVX0', 'CVX123']);\n",
```
Comment (ant0nsc) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:133]: these hardcoded classlabels should not be in master

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+    "import numpy as np\n",
+    "# plt.hist(df.model_output)\n",
+    "def logit(x):\n",
+    "    return np.log(x) - np.log(1.-x)\n",
+    "crossval_split = LoggingColumns.CrossValidationSplitIndex.value\n",
+    "n_splits = len(df[crossval_split].unique())\n",
+    "_, axs = plt.subplots(1, n_splits, figsize=(15, 5), sharex=True, sharey=True)\n",
+    "for (fold_index, fold_df), ax in zip(df.groupby(crossval_split), axs.flat):\n",
+    "    ax.hist([fold_df[fold_df.label==0].model_output, fold_df[fold_df.label==1].model_output], bins=16,\n",
+    "#     ax.hist([logit(fold_df[fold_df.label==0].model_output), logit(fold_df[fold_df.label==1].model_output)], bins=16,#np.linspace(-6, 6, 21),\n",
```
Comment (ant0nsc) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:152]: commented code?

File: InnerEye/ML/reports/classification_report.py [hunk 5]
```diff
@@ -65,7 +65,8 @@ def get_results(csv: Path) -> LabelsAndPredictions:
     return LabelsAndPredictions(subject_ids=subjects, labels=labels, model_outputs=model_outputs)
 
 
-def plot_auc(x_values: np.ndarray, y_values: np.ndarray, title: str, ax: Axes, print_coords: bool = False) -> None:
+def plot_auc(x_values: np.ndarray, y_values: np.ndarray, title: str, ax: Axes, print_coords: bool = False,
+             **plot_kwargs: Any) -> None:
```
Comment (ant0nsc) [inline InnerEye/ML/reports/classification_report.py:69]: can you add docu for plot_kwargs?

File: InnerEye/ML/reports/classification_report.py [hunk 6]
```diff
@@ -91,15 +92,24 @@ def plot_pr_and_roc_curves_from_csv(metrics_csv: Path) -> None:
...
-    fpr, tpr, thresholds = roc_curve(results.labels, results.model_outputs)
-    plot_auc(fpr, tpr, "ROC Curve", ax[0])
-    precision, recall, thresholds = precision_recall_curve(results.labels, results.model_outputs)
-    plot_auc(recall, precision, "PR Curve", ax[1])
 
-    plt.show()
+def plot_pr_and_roc_curves(labels: np.ndarray, model_outputs: np.ndarray, axs: Optional[Sequence[Axes]] = None,
+                           plot_kwargs: Optional[Dict[str, Any]] = None) -> None:
+    """
+    Plot the ROC and PR curves for the given predicted values and ground truth labels.
```
Comment (ant0nsc) [inline InnerEye/ML/reports/classification_report.py:102]: docu for all arguments please.

**Review by ant0nsc — CHANGES_REQUESTED** [review]

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "pycharm": {
+     "name": "#%%\n"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "df"
```
Comment (Shruthi42) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:71]: Can we remove this?

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "dataset = pd.read_csv(data_folder + \"dataset.csv\")\n",
+    "dataset"
```
Comment (Shruthi42) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:81]: We should remove the line that prints out the dataframe.

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+   "outputs": [],
+   "source": [
+    "from scipy.interpolate import interp1d\n",
+    "\n",
+    "def quantile(x, q, axis=-1):\n",
+    "    x = np.sort(x, axis=axis)\n",
+    "    rank = np.linspace(0, 1, x.shape[axis])\n",
+    "    return interp1d(rank, x, axis=axis)(q)\n",
+    "\n",
+    "x = np.random.randn(3, 4)\n",
```
Comment (Shruthi42) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:172]: Is this a test? It should go into a file in Tests\ML\reports.

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+    "sns.catplot(data=df, y='model_output', x='label', col='cross_validation_split_index', kind='box', aspect=.5);"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "plt.hist([df[df.label==0].model_output, df[df.label==1].model_output],\n",
```
Comment (Shruthi42) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:130]: This plot is missing a title.

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+    "df_.groupby('cross_validation_split_index')['subject'].nunique()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pd.crosstab(df_.cross_validation_split_index, [df_.label, df_.class_name], margins=True)"
```
Comment (Shruthi42) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:110]: This table is missing a title/explanation.

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+   "file_extension": ".py",
+   "mimetype": "text/x-python",
+   "name": "python",
+   "nbconvert_exporter": "python",
+   "pygments_lexer": "ipython3",
+   "version": "3.7.3"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 4
```
Comment (Shruthi42) [inline InnerEye/ML/reports/classification_crossval_report.ipynb:263]: Can we add a test to make sure this notebook runs correctly? We use nbformat version 4.5, which requires that each cell has a cell id, looks like that is not the case here.

**Review by Shruthi42 — CHANGES_REQUESTED** [review]

**Comment by melanibe** [comment]
I think we should test it on AML before merging this

File: InnerEye/ML/reports/classification_report.py [hunk 13]
```diff
@@ -52,84 +54,220 @@ class ReportedMetrics(Enum):
...
+    h2, = ax.plot(x_grid, y_mid, 'k', lw=2)
+    summary_handle = (h1, h2)
+    return line_handles, summary_handle
+
+
+def plot_pr_and_roc_curves_crossval(all_labels_and_model_outputs: Sequence[LabelsAndPredictions],
+                                    axs: Optional[Sequence[Axes]] = None) -> None:
+    """
+    Given a list of LabelsAndPredictions objects, plot the corresponding ROC and PR curves, along with median line and
+    shaded 80% confidence interval (computed over TPRs and precisions for each fixed FPR and recall value).
```
Comment (melanibe) [inline InnerEye/ML/reports/classification_report.py:214]: Why 80%?

File: InnerEye/ML/reports/classification_report.py [hunk 14]
```diff
@@ -52,84 +54,220 @@ class ReportedMetrics(Enum):
...
+                            ax: Optional[Axes] = None) -> Tuple[List, Any]:
+    """
+    Plot a collection of score curves along with the (vertical) median and confidence interval (CI).
+
+    Each plotted curve is interpolated onto a common horizontal grid, and the median and CI are computed vertically
+    at each horizontal location.
+    :param all_labels_and_model_outputs: Collection of ground-truth labels and model predictions (e.g. for various
+    cross-validation runs).
+    :param scoring_fn: A scoring function mapping a `LabelsAndPredictions` object to X and Y coordinates for plotting.
+    :param confidence_interval_width: A value in [0, 1] representing what fraction of the data should be contained in
```
Comment (melanibe) [inline InnerEye/ML/reports/classification_report.py:184]: To me it's weird to call that a confidence interval. If anything CI at 95% should be ~[mean +/- 2*std]

File: InnerEye/ML/reports/classification_report.py [hunk 15]
```diff
@@ -52,84 +54,220 @@ class ReportedMetrics(Enum):
...
+    confidence_interval_width = .8
+    line_handles, summary_handle = plot_scores_and_summary(all_labels_and_model_outputs,
+                                                           scoring_fn=get_roc_xy, ax=axs[0],
+                                                           confidence_interval_width=confidence_interval_width)
+    plot_scores_and_summary(all_labels_and_model_outputs, scoring_fn=get_pr_xy, ax=axs[1],
+                            confidence_interval_width=confidence_interval_width)
+
+    line_labels = [f"Split {split_index}" for split_index in range(len(all_labels_and_model_outputs))]
+    axs[0].legend(line_handles + [summary_handle],
+                  line_labels + [f"Median, {100 * confidence_interval_width:g}% CI"])
```
Comment (melanibe) [inline InnerEye/ML/reports/classification_report.py:241]: Same comment, from a statistical point of view it's not a CI. Alternatively you could say you plot the interquartile range as shaded area

File: InnerEye/ML/reports/classification_report.py [hunk 16]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
     """
     Given csvs written during inference for the validation and test sets, print out metrics for every prediction target
     in the config.
 
     :param val_metrics_csv: Csv written during inference time for the val set. This is used to determine the
-    optimal threshold for classification. 
+    optimal threshold for classification.
     :param test_metrics_csv: Csv written during inference time for the test set. Metrics are calculated for this csv.
     :param config: Model config
+    :param val_data_split: Whether to filter the validation CSV file for Train, Val, or Test results (default: no
```
[the remark this answers is not in the document]
Comment (melanibe) [reply InnerEye/ML/reports/classification_report.py:402]: Alternatively, if val_metrics may contain data from several split then we really need to change the name (cf. comment above)

File: InnerEye/ML/reports/classification_report.py [hunk 17]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
+                                                                 crossval_split_index=crossval_split,
+                                                                 data_split=test_data_split)
+        return get_all_metrics(val_labels_and_predictions, test_labels_and_predictions, is_thresholded)
 
     for prediction_target in config.class_names:
-        print_header(f"Class {prediction_target}", level=3)
-        val_metrics = get_labels_and_predictions(val_metrics_csv, prediction_target)
-        test_metrics = get_labels_and_predictions(test_metrics_csv, prediction_target)
-        print_metrics(val_labels_and_predictions=val_metrics, test_labels_and_predictions=test_metrics, is_thresholded=is_thresholded)
+        print_header(f"Class: {prediction_target}", level=3)
```
Comment (melanibe) [inline InnerEye/ML/reports/classification_report.py:420]: This is quite hard to parse, would you mind adding a few comments telling what each block is doing?

File: InnerEye/ML/reports/classification_report.py [hunk 18]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
+    set.
+    :param is_thresholded: Whether the model outputs are binary (they have been thresholded at some point)
+                           or are floating point numbers.
+    """
+    metrics = get_all_metrics(val_labels_and_predictions, test_labels_and_predictions, is_thresholded)
+    rows = [[description, f"{value:.4f}"] for description, value in metrics.items()]
+    print_table(rows)
 
 
 def print_metrics_for_all_prediction_targets(val_metrics_csv: Path,
```
Comment (melanibe) [inline InnerEye/ML/reports/classification_report.py:387]: Any way we can test this to make sure the aggregation of metrics is as expected? I didn't see a test

File: Tests/ML/reports/test_classification_report.py [hunk 19]
```diff
@@ -66,6 +68,71 @@ def test_generate_classification_report(test_output_dirs: OutputFolderForTests)
...
+    np.save(str(Path(config.local_dataset / f"1_{image_file_name}")),
+            np.random.randint(0, 255, [5, 4]))
+
+    result_file = test_output_dirs.root_dir / "report.ipynb"
+    result_html = generate_classification_crossval_notebook(result_notebook=result_file,
+                                                            config=config,
+                                                            crossval_metrics=crossval_metrics_file)
+    assert result_file.is_file()
+    assert result_html.is_file()
+    assert result_html.suffix == ".html"
```
Comment (melanibe) [inline Tests/ML/reports/test_classification_report.py:109]: Once we merge PR #432 we can also add a check that the AML classification cross-validation run, generated the report html file.

**Review by melanibe — CHANGES_REQUESTED** [review]

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 20]
```diff
@@ -0,0 +1,272 @@
...
+   "source": [
+    "# Default parameter values. They will be overwritten by papermill notebook parameters.\n",
+    "# This cell must carry the tag \"parameters\" in its metadata.\n",
+    "from pathlib import Path\n",
+    "import pickle\n",
+    "import codecs\n",
+    "\n",
+    "innereye_path = Path.cwd().parent.parent.parent\n",
+    "train_metrics_csv = \"\"\n",
+    "val_metrics_csv = innereye_path / 'Tests' / 'ML' / 'reports' / 'val_metrics_classification.csv'\n",
```
[the remark this answers is not in the document]
Comment (ant0nsc) [reply InnerEye/ML/reports/classification_crossval_report.ipynb:36]: For playing with the notebooks when developing them, it is incredibly helpful to have actual data here.

File: InnerEye/ML/reports/classification_report.py [hunk 21]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
+    :param val_data_split: Whether to filter the validation CSV file for Train, Val, or Test results (default: no
+    filtering).
+    :param test_data_split: Whether to filter the test CSV file for Train, Val, or Test results (default: no filtering).
     :param is_thresholded: Whether the model outputs are binary (they have been thresholded at some point)
                            or are floating point numbers.
-    """
+    :param is_crossval_report: If True, assumes CSVs contain results for multiple cross-validation runs and prints the
+    metrics along with means and standard deviations. Otherwise, prints metrics for a single run.
+    """
+    def get_metrics_for_fold(prediction_target: str, crossval_split: Optional[int] = None) -> Dict[str, float]:
```
Comment (ant0nsc) [inline InnerEye/ML/reports/classification_report.py:410]: based on Kristina's confusion about what a "fold" is, maybe we should start calling that a "split" or "cv_split".

File: InnerEye/ML/reports/classification_crossval_report.ipynb [hunk 22]
```diff
@@ -0,0 +1,272 @@
...
+    "test_metrics_csv = innereye_path / 'Tests' / 'ML' / 'reports' / 'test_metrics_classification.csv'\n",
+    "number_best_and_worst_performing = 20\n",
+    "config = \"\"\n",
+    "is_crossval_report = False"
```
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_crossval_report.ipynb:40]: what do you mean? we pass in `"is_crossval_report": False` for [standard reports](https://github.com/microsoft/InnerEye-DeepLearning/blob/2c296c7ae4e5cab151315ca41d41c43c237c2163/InnerEye/ML/reports/notebook_report.py#L156) and `True` for [cross-validation](https://github.com/microsoft/InnerEye-DeepLearning/blob/2c296c7ae4e5cab151315ca41d41c43c237c2163/InnerEye/ML/reports/notebook_report.py#L179)

File: InnerEye/ML/reports/notebook_report.py [hunk 23]
```diff
@@ -56,6 +56,25 @@ def print_header(message: str, level: int = 2) -> None:
     display(Markdown(f"{prefix} {message}"))
 
 
+def print_table(rows: Sequence[Sequence[str]], header: Optional[Sequence[str]] = None) -> None:
```
Comment (ant0nsc) [inline InnerEye/ML/reports/notebook_report.py:59]: Effectively you are creating a table from a dataframe here. I think there's already a built-in way of doing that with notebooks?

**Review by ant0nsc — CHANGES_REQUESTED** [review]

File: InnerEye/ML/reports/classification_report.py [hunk 24]
```diff
@@ -52,84 +54,220 @@ class ReportedMetrics(Enum):
     FalseNegativeRate = "false_negative_rate"
 
 
-def read_csv_and_filter_prediction_target(csv: Path, prediction_target: str) -> pd.DataFrame:
+def quantile(data: Any, quantiles: Any, axis: int = -1) -> np.ndarray:
```
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:57]: good point, my bad – I had in mind an older version of numpy where `np.percentile` had no `axis` argument. I'll replace it

File: InnerEye/ML/reports/classification_report.py (hunk 14 above)
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:184]: I disagree 🙂 strictly speaking, an X% CI is any interval that contains X% of the probability mass. it just happens that 95% is _mean +/- 1.96*std_ for a normal distribution. for these ROC/PR curves I don't think a normal assumption is appropriate (e.g. it's bounded in [0, 1]), so I went with quantiles.

File: InnerEye/ML/reports/classification_report.py (hunk 13 above)
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:214]: this choice was somewhat arbitrary. I thought 95% didn't make sense if we have only ~5–10 curves, and 50% (IQR) could be too narrow. but happy to take suggestions if people have a better idea

File: InnerEye/ML/reports/classification_report.py (hunk 15 above)
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:241]: (see replies above)

File: InnerEye/ML/reports/classification_report.py [hunk 25]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
-                                             is_thresholded: bool = False) -> None:
+                                             val_data_split: Optional[ModelExecutionMode] = None,
+                                             test_data_split: Optional[ModelExecutionMode] = None,
+                                             is_thresholded: bool = False,
+                                             is_crossval_report: bool = False) -> None:
     """
     Given csvs written during inference for the validation and test sets, print out metrics for every prediction target
     in the config.
 
     :param val_metrics_csv: Csv written during inference time for the val set. This is used to determine the
```
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:398]: yes, completely agree; I like your suggestion 👍

File: InnerEye/ML/reports/classification_report.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:402]: went for a name change (above), because for crossval runs we can use the `MetricsAcrossAllRuns.csv`, which contains val and test results

File: InnerEye/ML/reports/classification_report.py (hunk 17 above)
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:420]: sure, done

File: InnerEye/ML/reports/classification_report.py [hunk 26]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
+        print_header(f"Class: {prediction_target}", level=3)
+        all_metrics = []
+        if is_crossval_report:
+            header = ["Metric"]
+            for crossval_split in range(config.number_of_cross_validation_splits):
+                all_metrics.append(get_metrics_for_fold(prediction_target, crossval_split))
+                header.append(f"Split {crossval_split}")
+        else:
+            all_metrics.append(get_metrics_for_fold(prediction_target))
+            header = None  # type: ignore
```
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:429]: I felt a header was not really needed for this case, but now changed it to `["Metric", "Value"]` for single runs

File: InnerEye/ML/reports/classification_report.py [hunk 27]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
+        rows = [[metric] + [f"{fold_metrics[metric]:.4f}" for fold_metrics in all_metrics]
+                for metric in all_metrics[0]]
```
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:431]: done

File: InnerEye/ML/reports/classification_report.py [hunk 28]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
+        if is_crossval_report:
+            for row, metric in zip(rows, all_metrics[0]):
+                values = [fold_metrics[metric] for fold_metrics in all_metrics]
+                row.append(f"{np.mean(values):.4f} ({np.std(values):.4f})")
+            header.append("Mean (std)")
```
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:436]: decided to leave it here together with the creation of the aggregation column

File: InnerEye/ML/reports/classification_report.py (hunk 18 above)
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:387]: you mean the calculations of the metrics from the right CSV subsets, or the "Mean (std)" column?

File: InnerEye/ML/reports/classification_report.py [hunk 29]
```diff
@@ -176,92 +314,138 @@ def get_metric(val_labels_and_predictions: LabelsAndPredictions,
...
                            or are floating point numbers.
-    :return:
+    :return: Dictionary mapping metric descriptions to computed values.
     """
-
     optimal_threshold = 0.5 if is_thresholded else None
 
+    metrics = {}
     if not is_thresholded:
         roc_auc = get_metric(val_labels_and_predictions=val_labels_and_predictions,
```
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:334]: incorporated both suggestions, adding a `requires_threshold` attribute to determine which metrics should be included according to `is_thresholded`

File: InnerEye/ML/reports/classification_report.py (hunk 21 above)
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:378]: good catch, missed this one

File: InnerEye/ML/reports/notebook_report.py (hunk 23 above)
Comment (dccastro) [reply InnerEye/ML/reports/notebook_report.py:59]: yes, good point. looks a bit nicer, too 🙂

File: InnerEye/ML/reports/classification_report.py (hunk 14 above)
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py:173]: I'll just rename it 'interval' to avoid confusion

Comment (melanibe) [reply InnerEye/ML/reports/classification_report.py]: Yes but here you have a sample (of size n=number_splits) of the distribution over estimators not the true distribution so strictly speaking it's not a CI . But, Yes 1.96 assumues normal i agree not always respected for sure.

File: Tests/ML/reports/test_classification_report.py (hunk 19 above)
Comment (dccastro) [reply Tests/ML/reports/test_classification_report.py:87]: added check in `Tests/AfterTraining/test_after_training.py > test_expected_cv_files_classification()`

File: InnerEye/ML/reports/classification_report.py [hunk 30]
```diff
@@ -52,49 +53,80 @@ class ReportedScalarMetrics(Enum):
...
+        check_column_present(df, LoggingColumns.DataSplit)
+        df = df[df[LoggingColumns.DataSplit.value] == data_split.value]
+
+    # Filter by epoch
+    if LoggingColumns.Epoch.value in df:
+        # In a FULL_METRICS_DATAFRAME_FILE, the epoch column will be BEST_EPOCH_FOLDER_NAME (string) for the Test split.
+        # Here we cast the whole column to integer, mapping BEST_EPOCH_FOLDER_NAME to -1.
+        epochs = df[LoggingColumns.Epoch.value].apply(lambda e: -1 if e == BEST_EPOCH_FOLDER_NAME else int(e))
+        if epoch is None:
+            epoch = epochs.max()  # Take last epoch if unspecified
```
Comment (melanibe) [inline InnerEye/ML/reports/classification_report.py:101]:
```suggestion
            last_epoch = epochs.max()  # Take last epoch if unspecified
```

File: InnerEye/ML/reports/classification_report.py [hunk 31]
```diff
@@ -52,49 +53,80 @@ class ReportedScalarMetrics(Enum):
...
+        df = df[epochs == epoch]
```
Comment (melanibe) [inline InnerEye/ML/reports/classification_report.py:102]:
```suggestion
        df = df[epochs == last_epoch]
```

File: InnerEye/ML/reports/classification_crossval_report.ipynb (hunk 22 above)
[the remark this answers is not in the document]
Comment (melanibe) [reply InnerEye/ML/reports/classification_crossval_report.ipynb:40]: But it is two different notebooks so this param does not need to be passed to the notebook at all. If it is standard report you will use another ipynb file no?

File: InnerEye/ML/reports/classification_report.py (hunk 18 above)
Comment (melanibe) [reply InnerEye/ML/reports/classification_report.py]: I meant mean(std)

File: InnerEye/ML/reports/classification_crossval_report.ipynb (hunk 22 above)
[the remark this answers is not in the document]
Comment (dccastro) [reply InnerEye/ML/reports/classification_crossval_report.ipynb:40]: now we use `classification_crossval_report.ipynb` as the only template for both standard and crossval reports, and this switch controls what gets displayed. should I delete the legacy template file `classification_report.ipynb` to avoid confusion?

File: InnerEye/ML/reports/classification_report.py (hunk 18 above)
Comment (dccastro) [reply InnerEye/ML/reports/classification_report.py]: done

**Review by ant0nsc — DISMISSED** [review]

**Review by melanibe — APPROVED** [review]

**Review by melanibe — APPROVED** [review]

**Review by ant0nsc — APPROVED** [review]
